include/eve/product_type.hpp

```diff
@@ -69,4 +69,122 @@ namespace eve
+  std::bool_constant<std::same_as<Self, typename std::remove_cvref_t<Wrapper>::is_like>>
+  {};
+
+  template<typename Wrapper, typename Self>
+  concept like = std::same_as<std::remove_cvref_t<Wrapper>, Self> ||
+                 supports_like<std::remove_cvref_t<Wrapper>, Self>::value;
+
+  //================================================================================================
+  //! @addtogroup struct
+  //! @{
+  //!   @struct product_type_base<Self, ...Fields>
+  //!   @brief  a helper CRTP base to declare user defined types easier
+  //!
+  //!   It will generate a member wise equality.
+  //!   It will also generate a lexicographical order for your type, unless !supports_ordering_v<Self>
+  //!
+  //!   It generated the not mutating operator version from a mutating one (like operator+ from operator+=),
+  //!   however does not go further than that (like operator-= b from operator+= and operator-)
+  //!   since simd has an instruction for `-` and you want `operator-=` to be declared separately.
+  //!
+  //! @}
+  template <typename Self, typename ... Fields>
+  struct product_type_base;
+
+  namespace detail
+  {
+    std::false_type derived_from_product_type_base_impl(...);
+
+    template <typename Self, typename ... Fields>
+    std::true_type derived_from_product_type_base_impl(product_type_base<Self, Fields...> const&);
+
+    template <typename Type>
+    concept derived_from_product_type_base = decltype(derived_from_product_type_base_impl(Type{}))::value;
+  }
+
+  template <typename Self, typename ... Fields>
+  struct product_type_base : kumi::tuple<Fields...>
+  {
+    using tuple = kumi::tuple<Fields...>;
+
+    EVE_FORCEINLINE friend auto operator+(eve::like<Self> auto x, eve::like<Self> auto y) requires requires { x += y; }
+    {
+      x += y;
+      return x;
+    }
+
+    EVE_FORCEINLINE friend auto operator-(eve::like<Self> auto x, eve::like<Self> auto y) requires requires { x -= y; }
+    {
+      x -= y;
+      return x;
+    }
+
+    EVE_FORCEINLINE friend auto operator*(eve::like<Self> auto x, eve::like<Self> auto y) requires requires { x *= y; }
+    {
+      x *= y;
+      return x;
+    }
+
+    EVE_FORCEINLINE friend auto operator/(eve::like<Self> auto x, eve::like<Self> auto y) requires requires { x /= y; }
+    {
+      x /= y;
+      return x;
+    }
+
+    EVE_FORCEINLINE friend auto operator%(eve::like<Self> auto x, eve::like<Self> auto y) requires requires { x %= y; }
+    {
+      x %= y;
+      return x;
+    }
+
+    EVE_FORCEINLINE friend auto operator<<(eve::like<Self> auto x, integral_value auto s) requires requires { x <<= s; }
+    {
+      x <<= s;
+      return x;
+    }
+
+    EVE_FORCEINLINE friend auto operator>>(eve::like<Self> auto x, integral_value auto s) requires requires { x >>= s; }
+    {
+      x >>= s;
+      return x;
+    }
+
+    friend auto operator< (eve::like<Self> auto x, eve::like<Self> auto y) requires (!supports_ordering_v<Self>) = delete;
```

Comment:
Why do we need this ? IS it for hiding the ordering we inherit from kumi::tuple ?

Reply:
Miss the bitwise one

Reply:
> Why do we need this ? IS it for hiding the ordering we inherit from kumi::tuple ?

Yes. We want to use ordering from tuple. Unless oreding is disabled.

Will do the bitwise ones.